### PR TITLE
Trunk-5064: Add tests to PersonServiceImpl

### PR DESCRIPTION
**Pull request by ILauerIV**
## Description

Added Tests in relation to PersonServiceImpl
## Related Issue

see https://issues.openmrs.org/browse/TRUNK-5064

## Checklist:

- [X ] My pull request only contains one single commit.
- [ X] My pull request is based on the latest master branch
  `git pull --rebase upstream master`.
- [ ] I ran `mvn clean package` right before creating this pull request and
  added all formatting changes to my commit.
- [ ] My code follows the code style of this project.
- [ X] I have read the **CONTRIBUTING** document.
- [ X] I have added tests to cover my changes.
- [ X] All new and existing tests passed.

### CODE REVIEW

**Comment by teleivo** [comment]
@ILauerIV thanks for your contribution!!

* please take the time to read all of https://wiki.openmrs.org/display/docs/Pull+Request+Tips

* can you please check that your IDE settings are according to this [guide](https://wiki.openmrs.org/display/docs/Coding+Conventions#CodingConventions-IDE-configHowToConfigureYourIDE) your formatting is off

* please run the [tests locally](https://github.com/openmrs/openmrs-core#build-command) first to see if your changes work. and if your PR causes the build to fail, see travis error above please check why it failed. this is your responsibility.

* you are using the old style of unit tests. please check https://wiki.openmrs.org/display/docs/Generate+Test+Case+Plugin

**Comment by teleivo** [comment]
@ILauerIV please read the docs I posted in my previous comment, follow the tips and check your changes locally before you push so they conform with whats explained in the docs. if you have specific questions or problems just post here in a comment or ask on http://talk.openmrs.org/ and you'll get help :)

**Comment by teleivo** [comment]
@ILauerIV are you still working on this?

**Comment by lluismf** [comment]
There is already a PersonServiceTest, I don't see the point of having also a PersonServiceImplTest.

File: api/src/test/java/org/openmrs/api/PersonServiceTest.java [hunk 1]
```diff
@@ -2145,9 +2157,54 @@ public void savePersonAttributeType_shouldThrowAnErrorWhenTryingToSavePersonAttr
...
+		assertEquals("health", pat.getRetireReason());
+		assertEquals(new Date(), pat.getDateRetired());
+	}
+
 	
 	/**
 	 * @see PersonService#retirePersonAttributeType(PersonAttributeType, String)
+	 * @throws APIException
+	*/
+	@Test(expected = APIException.class)
```
Comment (teleivo) [inline api/src/test/java/org/openmrs/api/PersonServiceTest.java]: please use junits expected exception rule:
you can read about it here https://github.com/junit-team/junit4/wiki/Rules
 
and also find examples in openmrs-core 
* https://github.com/openmrs/openmrs-core/blob/a9a1fcee01ff6a8d40af6440be014a3b208ecc52/api/src/test/java/org/openmrs/DurationTest.java#L29-L30
* https://github.com/openmrs/openmrs-core/blob/a9a1fcee01ff6a8d40af6440be014a3b208ecc52/api/src/test/java/org/openmrs/DurationTest.java#L101-L102

this applies to all the tests you add which test for exceptions

File: api/src/main/java/org/openmrs/api/impl/PersonServiceImpl.java [hunk 2]
```diff
@@ -338,6 +338,7 @@ public Person savePerson(Person person) throws APIException {
 	
 	/**
 	 * @see org.openmrs.api.PersonService#voidPerson(org.openmrs.Person, java.lang.String)
+	 * @should Return Null When Person Is Null
```
Comment (ILauerIV) [inline api/src/main/java/org/openmrs/api/impl/PersonServiceImpl.java]: Can do.

File: api/src/main/java/org/openmrs/api/impl/PersonServiceImplTest.java [hunk 3]
```diff
@@ -0,0 +1,158 @@
+package org.openmrs.api.impl;
+
+import org.junit.Assert;
+
+/**
+ * Created by ira on 3/8/17.
+ */
+public class PersonServiceImplTest {
```
Comment (ILauerIV) [inline api/src/main/java/org/openmrs/api/impl/PersonServiceImplTest.java]: I am sorry about all of this. I am wrestling really hard wit Git and working across 2 computers. The auto generation tool created that extra class and I copied the code onto the main class for the real code. I am not sure  how the test code generation tool works.

**Comment by ILauerIV** [comment]
Sorry about the mess. I am having a large amount of trouble working with git and trying to both update the code to be line with the the master branch and squashing commits. The issue with the new empty class is that it was created by the testing coverage plugin when inserted the should commands. I copied the result of this program unto the test class. I think my problem was that I was not writing the @should on PersonService and instead on PersonServiceImpl.

**Comment by ILauerIV** [comment]
Sorry, I have been having trouble working with git and trying to both squash commits and rebase the branch. The issues with the should tags and the new empty class comes from my misunderstanding. I place the should tags in the PersonServiceImpl when I should have been placing them in PersonServiceTest. I was running the test generator and then copying that code over to PersonServiceTest from that empty class.

File: api/src/main/java/org/openmrs/api/impl/PersonServiceImpl.java (hunk 2 above)
Comment (ILauerIV) [inline api/src/main/java/org/openmrs/api/impl/PersonServiceImpl.java]: This problem caused my other 2 problems.

Comment (ILauerIV) [inline api/src/main/java/org/openmrs/api/impl/PersonServiceImpl.java]: Done

File: api/src/main/java/org/openmrs/api/impl/PersonServiceImplTest.java (hunk 3 above)
Comment (ILauerIV) [inline api/src/main/java/org/openmrs/api/impl/PersonServiceImplTest.java]: Done

File: api/src/main/java/org/openmrs/api/impl/PersonServiceImplTest.java [hunk 4]
```diff
@@ -0,0 +1,158 @@
...
+    /**
+     * @verifies Return Null When Person Is Null
+     * @see PersonServiceImpl#voidPerson(org.openmrs.Person, String)
+     */
+    @org.junit.Test
+    public void voidPerson_shouldReturnNullWhenPersonIsNull() throws Exception {
+        //TODO auto-generated
+        Assert.fail("Not yet implemented");
```
Comment (ILauerIV) [inline api/src/main/java/org/openmrs/api/impl/PersonServiceImplTest.java]: Removed.

File: api/src/test/java/org/openmrs/api/PersonServiceTest.java [hunk 5]
```diff
@@ -2187,4 +2244,216 @@ public void purgePersonAttributeType_shouldThrowAnErrorWhileTryingToDeletePerson
 		PersonAttributeType pat = ps.getPersonAttributeType(1);
 		ps.purgePersonAttributeType(pat);
 	}
+
+	//_____________________________________________________________________________________________________________________________-
```
Comment (ILauerIV) [inline api/src/test/java/org/openmrs/api/PersonServiceTest.java]: Done

File: api/src/test/java/org/openmrs/api/PersonServiceTest.java [hunk 6]
```diff
@@ -26,8 +26,8 @@
 import org.junit.Assert;
 import org.junit.Before;
 import org.junit.Test;
-import org.openmrs.GlobalProperty;
 import org.openmrs.Location;
+import org.openmrs.GlobalProperty;
```
Comment (ILauerIV) [inline api/src/test/java/org/openmrs/api/PersonServiceTest.java]: Thanks for the resource.

File: api/src/main/java/org/openmrs/api/impl/PersonServiceImpl.java (hunk 2 above)
Comment (ILauerIV) [inline api/src/main/java/org/openmrs/api/impl/PersonServiceImpl.java]: Done

File: api/src/main/java/org/openmrs/api/impl/PersonServiceImplTest.java (hunk 4 above)
Comment (ILauerIV) [inline api/src/main/java/org/openmrs/api/impl/PersonServiceImplTest.java]: Removed

File: api/src/test/java/org/openmrs/api/PersonServiceTest.java [hunk 7]
```diff
@@ -363,6 +364,17 @@ public void parsePersonName_shouldParseTwoPersonNameWithoutComma() throws Except
 		assertEquals("Doe", pname2.getFamilyName());
 		assertEquals("John", pname2.getGivenName());
 	}
+
+	@Test
+	@Verifies(value = "should set sort weight to one when all person attribute types are empty", method = "savePersonAttributeType(PersonAttributeType)")
+	public void saverPersonAttributeType_shouldSetSortWeightToOneWhenAllPersonAttributeTypesAreEmpty() throws Exception {
+		PersonService service = Context.getPersonService();
```
Comment (ILauerIV) [inline api/src/test/java/org/openmrs/api/PersonServiceTest.java]: I am working on this ticket with a class mate and he was in charge of this test method. I have asked him to explain this method.

File: api/src/test/java/org/openmrs/api/PersonServiceTest.java [hunk 8]
```diff
@@ -2187,4 +2244,216 @@ public void purgePersonAttributeType_shouldThrowAnErrorWhileTryingToDeletePerson
...
 		ps.purgePersonAttributeType(pat);
 	}
+
+	//_____________________________________________________________________________________________________________________________-
+
+	/**
+	 * @verifies Return Null When Person Is Null
+	 * @see PersonServiceImpl#voidPerson(org.openmrs.Person, String)
+	 */
+	@org.junit.Test
```
Comment (ILauerIV) [inline api/src/test/java/org/openmrs/api/PersonServiceTest.java]: Removed, these were auto generated in the PersonServiceImplTest and I just copied them over as such.

File: api/src/test/java/org/openmrs/api/PersonServiceTest.java [hunk 9]
```diff
@@ -2187,4 +2244,216 @@ public void purgePersonAttributeType_shouldThrowAnErrorWhileTryingToDeletePerson
...
+	}
+
+	/**
+	 * @verifies save a relationship type
+	 * @see PersonServiceImpl#saveRelationshipType(org.openmrs.RelationshipType)
+	 */
+	@org.junit.Test
+	public void saveRelationshipType_shouldSaveARelationshipType() throws Exception {
+		//TODO auto-generated
+		Assert.fail("Not yet implemented");
```
Comment (ILauerIV) [inline api/src/test/java/org/openmrs/api/PersonServiceTest.java]: Removed this method all together.

File: api/src/test/java/org/openmrs/api/PersonServiceTest.java.orig [hunk 10]
```diff
@@ -0,0 +1,2637 @@
+/**
```
Comment (ILauerIV) [inline api/src/test/java/org/openmrs/api/PersonServiceTest.java.orig]: Removed

**Comment by ILauerIV** [comment]
I have implemented most of the code comments with the exception of the method that my work mate is working on. I am still working on implementing the proper exception catching.  Thanks for the help. I will probably need more.

File: api/src/main/java/org/openmrs/api/PersonService.java [hunk 11]
```diff
@@ -124,6 +124,8 @@
 	 * @should set the date changed and changed by on update
 	 * @should update any global property which reference this type
 	 * @should throw an error when trying to save person attribute type while person attribute types are locked
+	 * @should set sort weight to one when all person attribute types are empty
+
```
Comment (teleivo) [inline api/src/main/java/org/openmrs/api/PersonService.java]: please remove empty line

File: api/src/main/java/org/openmrs/api/PersonService.java [hunk 12]
```diff
@@ -134,6 +136,9 @@
 	 * @param type
 	 * @param retiredReason
 	 * @should throw an error when trying to retire person attribute type while person attribute types are locked
+	 * @should person attribute type when retire reason is not null nor empty
+	 * @should  throw an error when retire reason is null
```
Comment (teleivo) [inline api/src/main/java/org/openmrs/api/PersonService.java]: remove extra whitespace

File: api/src/main/java/org/openmrs/api/PersonService.java [hunk 13]
```diff
@@ -639,6 +649,7 @@
 	 * @return Person person with given internal identifier
 	 * @throws APIException
 	 * @should return null when no person has the given id
+	 *  @should return null when personId is null
```
Comment (teleivo) [inline api/src/main/java/org/openmrs/api/PersonService.java]: remove indentation.

did you configure your IDE as described here
https://wiki.openmrs.org/display/docs/How-To+Setup+And+Use+Your+IDE
and ran the formatter (either via Eclipse save action or manually) ?

File: api/src/main/java/org/openmrs/api/impl/PersonServiceImpl.java [hunk 14]
```diff
@@ -193,6 +193,7 @@ public PersonAttributeType savePersonAttributeType(PersonAttributeType type) thr
 	
 	/**
 	 * @see org.openmrs.api.PersonService#retirePersonAttributeType(PersonAttributeType, String)
+
```
Comment (teleivo) [inline api/src/main/java/org/openmrs/api/impl/PersonServiceImpl.java]: please check your IDE formatter settings and then let the formatter take care of these issues here.

File: api/src/main/java/org/openmrs/api/impl/PersonServiceImpl.java [hunk 15]
```diff
@@ -477,6 +478,7 @@ public Relationship saveRelationship(Relationship relationship) throws APIExcept
 	/**
 	 * @see org.openmrs.api.PersonService#voidRelationship(org.openmrs.Relationship,
 	 *      java.lang.String)
+	 *
```
Comment (teleivo) [inline api/src/main/java/org/openmrs/api/impl/PersonServiceImpl.java]: same

File: api/src/main/java/org/openmrs/api/impl/PersonServiceImpl.java [hunk 16]
```diff
@@ -548,7 +550,7 @@ public void purgeRelationshipType(RelationshipType relationshipType) throws APIE
 	}
 	
 	/**
-	 * @see org.openmrs.api.PersonService#saveRelationshipType(org.openmrs.RelationshipType)
+	 * @see org.openmrs.api.PersonService#saveRelationshipType(org.openmrs.RelationshipType
```
Comment (teleivo) [inline api/src/main/java/org/openmrs/api/impl/PersonServiceImpl.java]: put the brace back, was probably an accident :)

File: api/src/main/java/org/openmrs/api/impl/PersonServiceImpl.java [hunk 17]
```diff
@@ -949,8 +951,9 @@ public RelationshipType retireRelationshipType(RelationshipType type, String ret
 		return Context.getPersonService().saveRelationshipType(type);
 	}
 	
-	/**
+	/**.
```
Comment (teleivo) [inline api/src/main/java/org/openmrs/api/impl/PersonServiceImpl.java]: also seems like an accident to add this dot here

File: api/src/main/java/org/openmrs/api/impl/PersonServiceImpl.java [hunk 18]
```diff
@@ -949,8 +951,9 @@ public RelationshipType retireRelationshipType(RelationshipType type, String ret
...
 	 * @see org.openmrs.api.PersonService#unretireRelationshipType(org.openmrs.RelationshipType)
+	 *
 	 */
 	@Override
 	public RelationshipType unretireRelationshipType(RelationshipType relationshipType) {
```
Comment (teleivo) [inline api/src/main/java/org/openmrs/api/impl/PersonServiceImpl.java]: this file should not show in your commit. remove all changes you made here please.

**Pull request closed without merging**